File: src/com/revbingo/spiff/vm/SpiffVm.java
/*******************************************************************************
 * Copyright 2012 Mark Piper
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package com.revbingo.spiff.vm;

import java.nio.ByteBuffer;
import java.util.List;

import com.revbingo.spiff.evaluator.Evaluator;
import com.revbingo.spiff.events.EventListener;
import com.revbingo.spiff.instructions.Instruction;
import com.revbingo.spiff.instructions.AdfInstruction;

public class SpiffVm {

	private List<Instruction> instructions;
	private ByteBuffer buffer;
	private EventListener ed;
	private Evaluator evaluator = new Evaluator();
	private int programCounter;
	private boolean isSuspended;
	private Object suspendedLock = new Object();

	public SpiffVm(List<Instruction> instructions, ByteBuffer buffer,
			EventListener ed) {
		this.instructions = instructions;
		this.buffer = buffer;
		this.ed = ed;

		evaluator.addVariable("fileLength", buffer.limit());
	}

	public void start() {
		for(Instruction i : instructions) {
			if(i instanceof AdfInstruction) {
				AdfInstruction vmi = (AdfInstruction) i;
				programCounter =vmi.lineNumber;

				if(vmi.isBreakpoint) {
					isSuspended = true;
					synchronized (suspendedLock) {
						try {
							suspendedLock.wait();
						} catch (InterruptedException e) {}
					}
				}
			}

			i.execute(buffer, ed, evaluator);
		}
	}

	public boolean isSuspended() {
		return isSuspended;
	}

	public Object getVar(String expression) {
		return evaluator.evaluate(expression);
	}

	public Integer getNextLineNumber() {
		return programCounter;
	}

	public void setBreakpoint(int i) {
		for(Instruction inst : instructions) {
			if(((AdfInstruction) inst).lineNumber == i) {
				((AdfInstruction) inst).isBreakpoint = true;
			}
		}
	}

	public void resume() {
		synchronized (suspendedLock) {
			suspendedLock.notify();
		}
	}
}

File: src/main/java/com/revbingo/spiff/vm/SpiffVm.kt
package com.revbingo.spiff.vm

import com.revbingo.spiff.datatypes.Datatype
import com.revbingo.spiff.evaluator.EvalExEvaluator
import com.revbingo.spiff.events.EventListener
import com.revbingo.spiff.instructions.AdfInstruction
import com.revbingo.spiff.instructions.Instruction
import java.nio.ByteBuffer

class SpiffVm(private val instructions: List<Instruction>, private val buffer: ByteBuffer,
              private val eventListener: EventListener = NullEventListener()) {

    private val evaluator = EvalExEvaluator()
    private val suspendedLock = Object()

    private var programCounter: Int = 0
    var isSuspended: Boolean = false
        private set

    val nextLineNumber: Int?
        get() = programCounter

    init {
        evaluator.addVariable("fileLength", buffer.limit())
    }

    fun start() {
        for (i in instructions) {
            if (i is AdfInstruction) {
                programCounter = i.lineNumber

                if (i.isBreakpoint) {
                    isSuspended = true
                    synchronized (suspendedLock) {
                        suspendedLock.wait()
                    }
                }
            }

            i.execute(buffer, eventListener, evaluator)
        }
    }

    fun getVar(expression: String): Any {
        return evaluator.evaluate(expression)
    }

    fun setBreakpoint(i: Int) {
        for (inst in instructions) {
            if(inst is AdfInstruction) {
                if (inst.lineNumber === i) {
                    inst.isBreakpoint = true
                }
            }
        }
    }

    fun resume() {
        synchronized (suspendedLock) {
            suspendedLock.notify()
        }
    }
}

class NullEventListener: EventListener {
    override fun notifyGroup(groupName: String, start: Boolean) {}
    override fun notifyData(ins: Datatype) {}
}